src/coreclr/vm/comutilnative.cpp

```diff
@@ -1681,226 +1681,137 @@ BOOL CanCompareBitsOrUseFastGetHashCode(MethodTable* mt)
-        hashCode ^= *pObj++;
-
-    return hashCode;
-}
-
-static INT32 RegularGetValueTypeHashCode(MethodTable *mt, void *pObjRef)
-{
-    CONTRACTL
-    {
+    CONTRACTL{
         THROWS;
         GC_TRIGGERS;
         MODE_COOPERATIVE;
     } CONTRACTL_END;
 
-    INT32 hashCode = 0;
+    // Should be handled by caller
+    _ASSERTE(!mt->CanCompareBitsOrUseFastGetHashCode());
 
-    GCPROTECT_BEGININTERIOR(pObjRef);
+    ValueTypeHashCodeStrategy ret = ValueTypeHashCodeStrategy::None;
 
-    BOOL canUseFastGetHashCodeHelper = FALSE;
-    if (mt->HasCheckedCanCompareBitsOrUseFastGetHashCode())
-    {
-        canUseFastGetHashCodeHelper = mt->CanCompareBitsOrUseFastGetHashCode();
-    }
-    else
-    {
-        canUseFastGetHashCodeHelper = CanCompareBitsOrUseFastGetHashCode(mt);
-    }
-
-    // While we should not get here directly from ValueTypeHelper::GetHashCode, if we recurse we need to
-    // be able to handle getting the hashcode for an embedded structure whose hashcode is computed by the fast path.
-    if (canUseFastGetHashCodeHelper)
-    {
-        hashCode = FastGetValueTypeHashCodeHelper(mt, pObjRef);
-    }
-    else
+    GCPROTECT_BEGININTERIOR(pObjRef);
+    
+    // it's looking ugly so we'll use the old behavior in managed code. Grab the first non-null
+    // field and return its hash code or 'it' as hash code
+    // <TODO> Note that the old behavior has already been broken for value types
+    //              that is qualified for CanUseFastGetHashCodeHelper. So maybe we should
+    //              change the implementation here to use all fields instead of just the 1st one.
+    // </TODO>
+    //
+    // <TODO> check this approximation - we may be losing exact type information </TODO>
+    ApproxFieldDescIterator fdIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
+
+    FieldDesc *field;
+    while ((field = fdIterator.Next()) != NULL)
     {
-        // it's looking ugly so we'll use the old behavior in managed code. Grab the first non-null
-        // field and return its hash code or 'it' as hash code
-        // <TODO> Note that the old behavior has already been broken for value types
-        //              that is qualified for CanUseFastGetHashCodeHelper. So maybe we should
-        //              change the implementation here to use all fields instead of just the 1st one.
-        // </TODO>
-        //
-        // <TODO> check this approximation - we may be losing exact type information </TODO>
-        ApproxFieldDescIterator fdIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
-
-        FieldDesc *field;
-        while ((field = fdIterator.Next()) != NULL)
+        _ASSERTE(!field->IsRVA());
+        if (field->IsObjRef())
         {
-            _ASSERTE(!field->IsRVA());
-            if (field->IsObjRef())
+            // if we get an object reference we get the hash code out of that
+            if (*(Object**)((BYTE *)pObjRef + field->GetOffsetUnsafe()) != NULL)
```

Comment:
```suggestion
            if (*(Object**)((BYTE *)objHandle->Unbox() + *fieldOffset + field->GetOffsetUnsafe()) != NULL)
```

Reply:
Should this statement be in cooperative mode? The unboxed pointer can be invalidated by other thread.

Reply:
Yes, it is the `GCX_COOP` from the other suggestion. GitHub did not let me to put both edits into one comment.